### PR TITLE
editoast: Utoipa for rollingstock

### PR DESCRIPTION
**Pull request by Tristramg**
Refs #4580

### CODE REVIEW

File: editoast/openapi.yaml [hunk 1]
```diff
@@ -4204,58 +4425,57 @@ paths:
...
+        required: true
+        schema:
+          format: int64
+          type: integer
+      - in: path
+        name: rolling_stock_id
         required: true
         schema:
+          format: int64
           type: integer
```
Comment (leovalais) [inline editoast/openapi.yaml]: duplicated parameter

File: editoast/src/views/rolling_stocks/mod.rs [hunk 2]
```diff
@@ -201,17 +257,29 @@ async fn update(
...
+    /// force the deletion even if it’s used
     #[serde(default)]
     force: bool,
 }
 
-#[delete("/{rolling_stock_id}")]
+/// Delete a rolling_stock and all entities linked to it
+#[utoipa::path(tag = "rolling_stock",
+    params(RollingStockIdParam),
+    request_body = DeleteRollingStockQueryParams,
```
Comment (leovalais) [inline editoast/src/views/rolling_stocks/mod.rs]:
```suggestion
    params(RollingStockIdParam, DeleteRollingStockQueryParams),
```

File: editoast/src/views/rolling_stocks/mod.rs [hunk 3]
```diff
@@ -58,21 +59,60 @@ pub enum RollingStockError {
     BasePowerClassEmpty,
 }
 
-pub fn routes() -> impl HttpServiceFactory {
-    scope("/rolling_stock")
-        .service((get_power_restrictions, get, create, update, delete))
-        .service(scope("/{rolling_stock_id}").service((create_livery, update_locked)))
+crate::routes! {
```
Comment (leovalais) [inline editoast/src/views/rolling_stocks/mod.rs:33]: Could you move this above `RollingStockError`?

File: editoast/src/views/rolling_stocks/mod.rs [hunk 4]
```diff
@@ -58,21 +59,60 @@ pub enum RollingStockError {
...
+                update_locked,
+            },
+            "/livery" => {
+                create_livery,
+            },
+        }
+    }
+}
+
+crate::schemas! {
```
Comment (leovalais) [inline editoast/src/views/rolling_stocks/mod.rs:53]: Likewise

File: editoast/src/models/rolling_stock/mod.rs [hunk 5]
```diff
@@ -47,53 +58,75 @@ pub struct RollingStockModel {
...
+    #[schema(value_type = f64)]
     pub mass: Option<f64>,
     #[diesel(deserialize_as = DieselJson<RollingResistance>)]
+    #[schema(value_type = RollingResistance)]
     pub rolling_resistance: Option<DieselJson<RollingResistance>>,
     #[diesel(deserialize_as = String)]
+    #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: Option<String>,
     #[diesel(deserialize_as = Option<JsonValue>)]
+    #[schema(value_type = Option<HashMap<String, String>>)]
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/mod.rs:110]: Wouldn't `models::train_schedule::PowerRestrictionRange` work here?

File: editoast/src/models/rolling_stock/mod.rs [hunk 6]
```diff
@@ -27,6 +29,14 @@ use editoast_derive::Model;
 use serde::{Deserialize, Serialize};
 use serde_json::Value as JsonValue;
 
+crate::schemas! {
+    RollingStock,
+    RollingStockWithLiveries,
+    crate::views::rolling_stocks::schemas(),
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/mod.rs]: We'd like schemas! to be a tree, could you include include these in `views/mod.rs`

File: editoast/src/schema/rolling_stock/mod.rs [hunk 7]
```diff
@@ -11,14 +11,32 @@ use crate::models::rolling_stock::rolling_stock_livery::RollingStockLiveryMetada
 
 crate::schemas! {
     RollingStockComfortType,
+    RollingStockCommon,
+    RollingStockWithLiveries,
+    RollingResistance,
+    RollingStockMetadata,
```
Comment (leovalais) [inline editoast/src/schema/rolling_stock/mod.rs:18]: Include this in models/rolling_stock/mod.rs

```suggestion
```

File: editoast/src/models/rolling_stock/light_rolling_stock.rs [hunk 8]
```diff
@@ -15,16 +15,19 @@ use diesel_json::Json as DieselJson;
 use editoast_derive::Model;
 use serde::Serialize;
 use serde_json::Value as JsonValue;
+use utoipa::ToSchema;
 
-#[derive(Debug, Model, Queryable, QueryableByName, Serialize)]
+#[derive(Debug, Model, Queryable, QueryableByName, Serialize, ToSchema)]
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/light_rolling_stock.rs:20]: If I'm not missing something this model is not included in any `schema!`

File: editoast/src/models/rolling_stock/light_rolling_stock.rs [hunk 9]
```diff
@@ -36,9 +39,13 @@ pub struct LightRollingStockModel {
     pub base_power_class: Option<String>,
     features: Vec<Option<String>>,
     mass: f64,
+    #[schema(value_type = RollingResistance)]
     rolling_resistance: DieselJson<RollingResistance>,
+    #[schema(value_type = LoadingGaugeType)]
     loading_gauge: String,
+    #[schema(value_type = HashMap<String, String>)]
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/light_rolling_stock.rs:46]: Ditto for `PowerRestrictionRanges`

File: editoast/src/schema/rolling_stock/light_rolling_stock.rs [hunk 10]
```diff
@@ -41,37 +51,43 @@ pub struct LightRollingStock {
...
+    #[schema(value_type = RollingResistance)]
     pub rolling_resistance: DieselJson<RollingResistance>,
     #[diesel(sql_type = Text)]
+    #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: String,
     #[diesel(sql_type = Jsonb)]
+    #[schema(value_type = RollingStockMetadata)]
     pub metadata: DieselJson<RollingStockMetadata>,
     #[diesel(sql_type = Nullable<Jsonb>)]
+    #[schema(value_type = HashMap<String, String>)]
```
Comment (leovalais) [inline editoast/src/schema/rolling_stock/light_rolling_stock.rs:63]: PowerRestrictionRanges

File: editoast/src/views/rolling_stocks/mod.rs [hunk 11]
```diff
@@ -58,21 +59,60 @@ pub enum RollingStockError {
...
+    RollingStockForm,
+    DeleteRollingStockQueryParams,
+    RollingStockLockedUpdateForm,
+    RollingStockLiveryCreateForm,
+    PowerRestriction,
+    RollingStockLivery,
```
Comment (leovalais) [inline editoast/src/views/rolling_stocks/mod.rs]: Include this in schema/rolling_stocks/mod.rs
```suggestion
```

File: editoast/src/schema/rolling_stock/mod.rs [hunk 12]
```diff
@@ -30,40 +48,48 @@ pub struct RollingStockCommon {
     pub features: Vec<String>,
     pub mass: f64,
     pub rolling_resistance: RollingResistance,
+    #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: String,
+    /// Mapping of power restriction code to power class
+    #[schema(value_type = HashMap<String, String>)]
```
Comment (leovalais) [inline editoast/src/schema/rolling_stock/mod.rs:55]: PowerRestrictionRange

File: front/src/common/api/osrdEditoastApi.ts [hunk 13]
```diff
@@ -1488,108 +1473,119 @@ export type ViewMetadata = {
...
+  soc_ref: number;
+  tau: number;
 };
 export type EnergyStorage = {
   capacity: number;
-  refill_law: {
-    soc_ref: number;
-    tau: number;
-  } | null;
+  refill_law?: RefillLaw | null;
```
Comment (leovalais) [inline front/src/common/api/osrdEditoastApi.ts]: Is this field really optional?

File: front/src/common/api/osrdEditoastApi.ts [hunk 14]
```diff
@@ -1488,108 +1473,119 @@ export type ViewMetadata = {
...
+  unit: string;
+};
+export type RollingResistance = {
+  A: number;
+  B: number;
+  C: number;
+  type: string;
+};
+export type LightRollingStock = {
+  base_power_class?: string | null;
```
Comment (leovalais) [inline front/src/common/api/osrdEditoastApi.ts:1553]: Can this field really be undefined?

File: front/src/common/api/osrdEditoastApi.ts [hunk 15]
```diff
@@ -1800,48 +1796,116 @@ export type ScenarioPatchForm = {
   name?: string | null;
   tags?: string[] | null;
 };
-export type Comfort = 'AC' | 'HEATING' | 'STANDARD';
+export type RollingStockComfortType = 'STANDARD' | 'AC' | 'HEATING';
+export type EffortCurveConditions = {
```
Comment (leovalais) [inline front/src/common/api/osrdEditoastApi.ts:1800]: These fields are probably never undefined

**Review by leovalais — CHANGES_REQUESTED** [review]
Thanks this one's massive! Here's a first batch of comments on issues I could pick up reading the changes. I'll do another pass on editoast changes once these are resolved.

File: editoast/src/models/rolling_stock/mod.rs (hunk 5 above)
Comment (Tristramg) [reply editoast/src/models/rolling_stock/mod.rs:110]: I don’t think so. The `PowerRestrictionRange` works on a range (between two floats), while here is it a string<>string mapping

File: front/src/common/api/osrdEditoastApi.ts (hunk 13 above)
Comment (Tristramg) [reply front/src/common/api/osrdEditoastApi.ts]: it is in the existing code
https://github.com/osrd-project/osrd/blob/dev/editoast/src/schema/rolling_stock/mod.rs#L185

File: front/src/common/api/osrdEditoastApi.ts (hunk 14 above)
Comment (Tristramg) [reply front/src/common/api/osrdEditoastApi.ts:1553]: https://github.com/osrd-project/osrd/blob/dev/editoast/src/schema/rolling_stock/mod.rs#L22

File: front/src/common/api/osrdEditoastApi.ts (hunk 15 above)
Comment (Tristramg) [reply front/src/common/api/osrdEditoastApi.ts:1800]: https://github.com/osrd-project/osrd/blob/dev/editoast/src/schema/rolling_stock/mod.rs#L107-L110 They are all optional here, should I override the schema to make them non optional?

File: editoast/src/models/rolling_stock/mod.rs (hunk 5 above)
Comment (leovalais) [reply editoast/src/models/rolling_stock/mod.rs:110]: You're right, my bad...

**Review by leovalais — APPROVED** [review]
Tested. THANKS!!! 😄

File: front/src/common/api/osrdEditoastApi.ts [hunk 16]
```diff
@@ -989,8 +985,7 @@ export type PostPathfindingApiArg = {
 };
 export type DeletePathfindingByIdApiResponse = unknown;
 export type DeletePathfindingByIdApiArg = {
-  /** Path ID */
-  id: number;
+  id?: number;
```
Comment (clarani) [inline front/src/common/api/osrdEditoastApi.ts]: Why this id is now optional ? I think it's a mistake, it should not become optional

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 17]
```diff
@@ -1,19 +1,26 @@
...
   buttonText?: string;
   deleteAction?: boolean;
 };
 
+type RollingStockIsUsed = {
+  RollingStockIsUsed: {
+    rolling_stock_id: number;
+    usage: TrainScheduleScenarioStudyProject[];
+  };
+};
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]: You can remove this, see in display function below

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 18]
```diff
@@ -25,8 +32,8 @@ const RollingStockEditorFormModal = ({
   const { closeModal } = useModal();
   const { t } = useTranslation(['translation', 'rollingstock']);
 
-  const displayErrorObject = (errorList: RollingStockUsage['usage'][]): JSX.Element => {
-    const projectList = groupBy(errorList, 'project_name');
+  const displayErrorObject = (errorList: RollingStockIsUsed): JSX.Element => {
+    const projectList = groupBy(errorList.RollingStockIsUsed, 'project_name');
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]: You can test if the rollingStockError is a RollingStockIsUsed error like this, this way no need to use a cast
```suggestion
  const displayErrorObject = (errorList: RollingStockError) => {
    if (typeof errorList !== 'object' || !('RollingStockIsUsed' in errorList)) {
      return null;
    }
    const projectList = groupBy(errorList.RollingStockIsUsed, 'project_name');
```

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 19]
```diff
@@ -52,7 +59,7 @@ const RollingStockEditorFormModal = ({
       {!errorObject ? (
         <span className="text-primary mb-3">{mainText}</span>
       ) : (
-        displayErrorObject(errorObject)
+        displayErrorObject(errorObject as RollingStockIsUsed)
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]: Remove the cast here
```suggestion
        displayErrorObject(errorObject)
```

File: front/src/modules/rollingStock/components/RollingStockSelector/WithRollingStockSelector.tsx [hunk 20]
```diff
@@ -11,33 +11,33 @@ const WithRollingStockSelector =
   (hocProps: T) => {
     const { t } = useTranslation(['translation', 'rollingstock']);
 
-    const [rollingStockSelected, setRollingStockSelected] = useState<RollingStock | undefined>(
-      undefined
-    );
-    const rollingStockID = useSelector(getRollingStockID);
+    const [rollingStockSelected, setRollingStockSelected] = useState<
+      RollingStockWithLiveries | undefined
+    >(undefined);
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockSelector/WithRollingStockSelector.tsx]: Only a small suggestion :) 
```suggestion
    >();
```

File: front/src/modules/rollingStock/components/RollingStockSelector/WithRollingStockSelector.tsx [hunk 21]
```diff
@@ -11,33 +11,33 @@ const WithRollingStockSelector =
...
+    const rollingStockId = useSelector(getRollingStockID);
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockSelector/WithRollingStockSelector.tsx:17]: Thanks for changing that :pray:

**Review by clarani — COMMENTED** [review]
Thanks for this very nice PR, I left some comments

File: front/src/common/api/osrdEditoastApi.ts (hunk 16 above)
Comment (Tristramg) [reply front/src/common/api/osrdEditoastApi.ts]: This is strange, I’m not sure what happened here. I’ll ask @leovalais

Comment (Tristramg) [reply front/src/common/api/osrdEditoastApi.ts]: Just a bad rebase :grimacing: 
It should be all right now

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 22]
```diff
@@ -1,19 +1,20 @@
...
   // request can be a POST, PUT, PATCH or DELETE request
   request?: () => void;
   mainText: string;
-  errorObject?: RollingStockUsage['usage'][];
+  errorObject?: RollingStockError;
   buttonText?: string;
   deleteAction?: boolean;
 };
 
+
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]:
```suggestion
```

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 23]
```diff
@@ -1,19 +1,20 @@
...
+import { RollingStockError, TrainScheduleScenarioStudyProject } from 'common/api/osrdEditoastApi';
 import { groupBy } from 'lodash';
 
 type RollingStockEditorFormModalProps = {
   setAddOrEditState?: React.Dispatch<React.SetStateAction<boolean>>;
   // request can be a POST, PUT, PATCH or DELETE request
   request?: () => void;
   mainText: string;
-  errorObject?: RollingStockUsage['usage'][];
+  errorObject?: RollingStockError;
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]:
```suggestion
  errorObject?: TrainScheduleScenarioStudyProject[];
```

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx [hunk 24]
```diff
@@ -25,8 +26,11 @@ const RollingStockEditorFormModal = ({
   const { closeModal } = useModal();
   const { t } = useTranslation(['translation', 'rollingstock']);
 
-  const displayErrorObject = (errorList: RollingStockUsage['usage'][]): JSX.Element => {
-    const projectList = groupBy(errorList, 'project_name');
+  const displayErrorObject = (errorList: RollingStockError) => {
+    if (typeof errorList !== 'object' || !('RollingStockIsUsed' in errorList)) {
+      return null;
+    }
+    const projectList = groupBy(errorList.RollingStockIsUsed, 'project_name');
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormModal.tsx]:
```suggestion
  const displayErrorObject = (errorList: TrainScheduleScenarioStudyProject[]) => {
    const projectList = groupBy(errorList, 'project_name');
```

**Review by clarani — CHANGES_REQUESTED** [review]
Tested, found a small bug and left last comments

**Review by clarani — APPROVED** [review]
LGTM and tested :white_check_mark:

**Pull request merged**